be/src/vec/runtime/vdatetime_value.cpp

```diff
@@ -3367,16 +3370,28 @@ void DateV2Value<T>::from_unixtime(int64_t timestamp, int32_t nano_seconds,
     static const cctz::time_point<cctz::sys_seconds> epoch =
             std::chrono::time_point_cast<cctz::sys_seconds>(
                     std::chrono::system_clock::from_time_t(0));
-    cctz::time_point<cctz::sys_seconds> t = epoch + cctz::seconds(timestamp);
-
-    const auto tp = cctz::convert(t, ctz);
 
-    if (scale > 6) [[unlikely]] {
+    // Do rounding before looking up the time zone.
+    if (scale > 6) {
         scale = 6;
     }
 
+    nano_seconds /= common::exp10_i32(9 - scale - 1);
+
+    if (nano_seconds % 10 >= 5) {
+        nano_seconds += 10;
+    }
+
+    if (nano_seconds >= common::exp10_i32(scale + 1)) {
+        timestamp += 1;
+        nano_seconds = 0;
+    }
+
+    int32_t millisecond = (nano_seconds /= 10) * common::exp10_i32(6 - scale);
```

Comment:
6 width precision is microsecond, not millisecond

Reply:
fixed